## Ασκήσεις 35-37


In [35]:
import random
random.seed(2018)  # for reproducibility
from simple_markov_chain_lib import markov_chain
import statistics as stat ## import the library statistics. We will use it to compute the mean and variance of our list


p = 1/6

# A dictionary for the initial distibution. 
# We prescribe the initial distribution
init_probs = {1: 1.0} 
 
# A dictionary for the transition probability  matrix. 
# Every state-key corresponds to a list with tuples of (Next_State,Probability) 
markov_table = {
    1: {2: 1.},
    2: {2: 2/3, 3: 1/3},
    3: {1: p, 2: 1-p}
}
 
# Ok... we are ready know
# Let's construct a Markov Chain. So let's call the constructor
mc = markov_chain(markov_table, init_probs)
estimates=[] ## create the empty list estimates.
M = 50

## Experiment parameters

for j in range(0,M):
    N = 200     # number of samples
    steps = 20   # the target time
    counter = 0  # to count the number of times the event {X_40  = 1} occurs

    ## Simulation
    for i in range(N):
        mc.start()  # new experiment
        for j in range(steps):  mc.move()
        if mc.running_state == 1:  counter += 1

    phat = counter / N
    estimates.append(phat)
    
print(
    """ 
    For M = 50, N = 200
    The sample mean is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(estimates), stat.variance(estimates))
)

estimates=[]

for j in range(0,M):
    N = 20000     # number of samples
    steps = 20   # the target time
    counter = 0  # to count the number of times the event {X_40  = 1} occurs

    ## Simulation
    for i in range(N):
        mc.start()  # new experiment
        for j in range(steps):  mc.move()
        if mc.running_state == 1:  counter += 1

    phat = counter / N
    estimates.append(phat)
    
print(
    """ 
    For M = 50, N = 20000
    The sample mean is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(estimates), stat.variance(estimates))
)


 
    For M = 50, N = 200
    The sample mean is 0.03890 and the sample variance is 0.00012
    
 
    For M = 50, N = 20000
    The sample mean is 0.03998 and the sample variance is 0.00000
    


In [36]:
import numpy as NP
from numpy.linalg import matrix_power
from scipy import linalg as LA
A = NP.array([[0, 1, 0],
           [0, 2/3, 1/3],
           [1/6, 5/6, 0]],dtype = NP.float64)

A = matrix_power(A, 20)
print(A)

[[0.04 0.72 0.24]
 [0.04 0.72 0.24]
 [0.04 0.72 0.24]]


## Απαντήσεις ανάλυσης

1. Ποια είναι ο μέσος όρος των $Μ$ εκτιμήσεων της πιθανότητας $\mathbb{P}_1[X_{20}=1]$ όταν $N=200$ και ποιος όταν $N=20.000$;

    Ο μέσος όρος όπως φαίνεται από το πρώτο μπλοκ κώδικα είναι για $200$ και $20.000$ δείγματα $0.0389$ και $0.03998$ αντίστοιχα.


2. Είναι αυτά τα αποτελέσματα κοντά στη θεωρητική τιμή που έχετε υπολογίσει;

    Η θεωρητική τιμή υπολογίζεται από τον πίνακα μεταβάσεων P υψωμένο στην 20η δύναμη. Προσεγγιστικά και μέσω python στο δεύτερο μπλοκ κώδικα υπολογίζεται ο πίνακας μετάβασης στην 20η δύναμη και προκύπτει ότι το αποτέλεσμα ειναι οριακά διαφορετικό της πραγματικής τιμής $\mathbb{p11}$ = $0.04$.
    
    
3. Ποια είναι η δειγματική διασπορά των $M$ εκτιμήσεων στις δύο περιπτώσεις;

    Η δειγματική διασπορά προκύπτει επίσης από το πρώτο μπλοκ κώδικα για $200$ και $20.000$ δείγματα ως $0.00012$ και $0.00000$.
    
    
4. Αλλάζει ουσιαστικά η εκτίμησή σας, αν η αλυσίδα ξεκινάει από την κατάσταση `3` (τρίτη γραμμή του πίνακα);

    Όχι, η πιθανότητα να βρεθείς σε μια κατάσταση έπειτα απο ικανοποιητικό αριθμό βημάτων εξαρτάται μονάχα από την κατάσταση στόχο και όχι απο την αρχική συνθήκη (όπως φαίνεται και από τα αποτελέσματα στο δεύτερο μπλοκ κώδικα). Αυτό δεν ισχύει εν γένει σε όλες τις μαρκοβιανές αλυσίδες.
    

## Άσκηση 40

In [1]:
import random
random.seed(2018)  # for reproducibility
from simple_markov_chain_lib import markov_chain
import statistics as stat ## import the library statistics. We will use it to compute the mean and variance of our list

# A dictionary for the initial distibution. 
# We prescribe the initial distribution
init_probs = {"0-0": 1.0} 
 
# A dictionary for the transition probability  matrix. 
# Every state-key corresponds to a list with tuples of (Next_State,Probability) 
p = 0.6
q = 1-p
markov_table = {
    "A": {"A": 1},
    "B": {"B": 1},
    
    "0-0": {"15-0": p, "0-15": q},
    "15-0": {"30-0": p, "15-15": q},
    "0-15": {"15-15": p, "0-30": q},
    "30-0": {"40-0": p, "30-15": q},
    "0-30": {"15-30": p, "0-40": q},
    "40-0": {"A": p, "40-15": q},
    "0-40": {"15-40": p, "B": q},

    "15-15": {"30-15": p, "15-30": q},
    "Deuce": {"AdvA": p, "AdvB": q},
    
    "30-15": {"40-15": p, "Deuce": q},
    "15-30": {"Deuce": p, "15-40": q},
    "40-15": {"A": p, "AdvA": q},
    "15-40": {"AdvB": p, "B": q},

    "AdvB": {"Deuce": p, "B": q},
    "AdvA": {"A": p, "Deuce": q}
    
}
 
# Ok... we are ready know
# Let's construct a Markov Chain. So let's call the constructor
mc = markov_chain(markov_table, init_probs)
estimates=[] ## create the empty list estimates.
M = 50

## Experiment parameters

for j in range(0,M):
    N = 50000     # number of samples
    counter = 0  # to count the number of times the event {X_40  = 1} occurs

    ## Simulation
    for i in range(N):
        mc.start()  # new experiment
        while mc.running_state != "A" and mc.running_state != "B":  mc.move()
        if mc.running_state == "A":  counter += 1
        
    phat = counter / N
    estimates.append(phat)
    
print(
    """ 
    For M = 50, N = 50000
    The sample mean is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(estimates), stat.variance(estimates))
)


 
    For M = 50, N = 50000
    The sample mean is 0.73621 and the sample variance is 0.00000
    


# Απαντήσεις ανάλυσης

Θεωρούμε την μαρκοβιανή αλυσίδα με καταστάσεις όλες τις πιθανές καταστάσεις ενος παιχνιδιού τένις (17 καταστάσεις). Η πιθανότητα μετάβασης από κάθε κατάσταση σε κάθε άλλη (εκτός των δύο τελικών καταστάσεων) επιλέγεται ως p = $0.6$ για πόντο του παίκτη Α και q = 1 - p = $0.4$ για πόντο του παίκτη Β. Προσομοιάζουμε το παιχνίδι μετρώντας τις φορές που νίκησε ο Α σε $50.000$ παιχνίδια 50 φορές. Από αυτές τις εκτιμήσεις λαμβάνουμε την δειγματική μέση τιμή και διασπορά της εκτίμησης πιθανότητας νίκης του Α ως $0.73621$ και $0$ αντίστοιχα!